### DISCUSSION TITLE
Nested Upserts

### DISCUSSION THREAD

**Discussion by RoboZoom** [zookzook/elixir-mongodb-driver#166] [Ideas]
Hi,

I wrote a utility for upserts of objects in nested arrays. Is this something you'd be interested in incorporating into the library?  If so, I'd be happy to contribute it!

**Comment by zookzook** [comment]
I think this is a topic for the Repo module? Could you provide an example?

**Reply by RoboZoom** [reply]
If you have a data structure in a collection "foos" like:

```json
{
    "foo": {
        "id": "1a",
        "bar": {
            "baz": [
                {
                    "id": "1",
                    "name": "First",
                    "description": "Some information"
                },
                {
                    "id": "2",
                    "name": "Second",
                    "description": "Some Additional Information"
                },
                {
                    "id": "3",
                    "name": "Third",
                    "description": "This is the third item"
                }
            ]
        }
    }
}
```

If you want to upsert into the `baz` array, there's no terse way of doing this with Mongo.  You either need to make it into a multiple operation type transaction (check of the new value is in the array - if so, edit the array... if not, push it to the array), or you can build an aggregation pipeline .  

I've implemented a general form of the aggregation pipeline which takes a key-path to the array, a key to match identity amongst the array items, and it executes a partial update against an element that already exists or pushes it into the array.

For example
```elixir
upsert_nested("foos", "1a", "foo.bar.baz", "id", "%{id: "3", name: "Third +")
```
Would change the last element to:
```json
{
                    "id": "3",
                    "name": "Third +",
                    "description": "This is the third item"
                }
```

I think it would need to be a little bit more general than it currently is (and I may need your help to best do that) but it's an 80% solution right now.

**Comment by zookzook** [comment]
It could be a nice extension, but it uses some kind of `id` which depends on the use case:
```
upsert_nested("foos", "1a", "foo.bar.baz", "id", "%{id: "3", name: "Third +")
```
This approach assumes that the data structure has a concept of id. But it is not clear, that the second parameter is the value of an attribute named `id`. 

I think the example is not correct: instead of 
```
{
    "foo": {
        "id": "1a",
        "bar": {
```

the document should look like:

```
{
    "id": "1a",
    "foo": {
        "bar": {
```
?

Someone would use "_id" or "uuid" as the identifier. Maybe the API can look like this:

```
upsert_one_nested(collection, query, path, subquery, doc)
````
The function will use the `query` to find the document in the `collection`, the `path` points to the nested documents. The `subquery` is used to find the nested document, which is updated to `doc` or when missing, it is inserted under the `path`.

```
upsert_one_nested("foos", %{id: "1a"}, "foo.bar.baz", %{id: "3"}, %{name: "Third +"})
```
